### ISSUE TITLE
Cannot Modify The Element From Document

### ISSUE THREAD

**Issue by shtse8** [Imangazaliev/DiDOM#69]
Sample Code:
```
$document = new Document('<div id="content"><div class="a">b</div><div class="a">c</div></div>');
$contentDom = $document->find('#content')[0];
$contentDom->find('.a')[0]->remove();
echo $contentDom->innerHtml();
```

Expected Result:
```
<div class="a">c</div>
```

Result:
```
<div class="a">b</div><div class="a">c</div>
```

**Comment by shtse8** [comment]
I can produce the nearest result for the following code: But Document hasn't innerHTML method and it will return <html><body> ... </body></html> which I don't need.
```
$document = new Document('<div id="content"><div class="a">b</div><div class="a">c</div></div>');
$contentDom = $document->find('#content')[0];
if ( $contentDom ) {
	$contentDom = new Document($contentDom->innerHtml());
	$contentDom->find('.a')[0]->remove();
	echo $contentDom->html();
}
```

I've tested only the Document can be modified.

**Comment by shtse8** [comment]
I am moving to DiDom from HtmlDomParser. HtmlDomParser can modify the Element and return the correct html by the following code:
```
$document = str_get_html('<div id="content"><div class="a">b</div><div class="a">c</div></div>');
$contentDom = $document->find('#content', 0);
$contentDom->find('.a', 0)->outertext= '';
echo $contentDom->innertext;
```

Result:
```
<div class="a">c</div>
```

**Comment by shtse8** [comment]
I can produce the expected result by the following code:
Using toDocument(), I don't need to remove the html and body tags by myself.
But I still need to filter out the parent Dom for the result.
```
$document = new Document('<div id="content"><div class="a">b</div><div class="a">c</div></div>');
$contentDom = $document->find('#content')[0];
if ( $contentDom ) {
	$contentDom = $contentDom->toDocument();
	$contentDom->first('.a')->remove();
	/* We want innerHTML but Document doesn't have, so we need to get the first element for calling innerHTML()  */
	echo $contentDom->first('*')->innerHtml();
}
```

Result:
```
<div class="a">c</div>
```

**Comment by Imangazaliev** [comment]
Hello. DiDom not allows modify elements founded in another element. You can find explanation here: #48, #51

**Comment by Imangazaliev** [comment]
I think that your second example (with converting to Document) it is the only way

**Comment by shtse8** [comment]
Yes, I understand the reason because I have read the source code which the find method of element create a new document. But I hope DiDom can support it soon? it's common to modify elements found in another element. For example, we want to modify the Form's elements only, and for the best practice, we should save the form element into a variable and then try to find in that element.

**Comment by shtse8** [comment]
The second example works. But if we want to get the full document finally (not only the contentDom), the full document is still the original one. And it's no way to merge the modified element to the original document. Since we cannot modify the element found in another element, if we need to modify lots of elements, it's a tough work.

**Comment by Imangazaliev** [comment]
I will think how to solve this problem

**Comment by shtse8** [comment]
Any Updates?

**Comment by Imangazaliev** [comment]
Oh, yes, i will release new version as soon as possible. Four or five days may be

**Comment by Imangazaliev** [comment]
Released: (1.9.0)[https://github.com/Imangazaliev/DiDOM/releases/tag/1.9.0]. Usage:

```php
$contentDom = $document->find('#content')[0];

$document->find('.a', Query::TYPE_CSS, true, $contentDom)[0]->remove();

echo $contentDom->innerHtml();
```

**Note:** you can use `first()` method instead of `find(...)[0]`.

**Comment by shtse8** [comment]
Thanks for your update.
For my best understanding, does it mean we have to modify anything from the document directly by passing $contentDom to the find method if we want to get the modified document finally?
Is it possible to use `find` from the $contentDom and make changes to the document? it should have better programming experience.

**Comment by Imangazaliev** [comment]
For my best understanding, does it mean we have to modify anything

Yes

    Is it possible to use find from the $contentDom and make changes to the document? 

Yes. i want to implement it in next version. I will create a separate method is likely (`findInDocument` for example). I can't change behavior of `find` method because search wit context node works incorrect for nodes without owner document. What you think about this.

**Comment by shtse8** [comment]
How about changing the `find` method to return the referenced document object storing the node xpath? then changing every method to work on that xpath. So developers don't need to pass the `$contentDom` into every find as the forth argument.

**Comment by Imangazaliev** [comment]
Released: [1.9.1](https://github.com/Imangazaliev/DiDOM/releases/tag/1.9.1). Usage:

```php
$contentDom = $document->find('#content')[0];

$contentDom->findInDocument('.a')[0]->remove();

echo $contentDom->innerHtml();
```

Also available `firstInDocument` method:

```php
$contentDom->firstInDocument('.a')->remove();
```
